pkg/backends/calico/routes.go

```diff
@@ -206,57 +194,114 @@ func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
 		routes := rg.getAdvertisedRoutes(key)
 		rg.withdrawRoutesForKey(key, routes)
 	}
-
 }
 
-// onBGPConfigurationUpdate is called from the calico client, whenever there
+// onExternalIPsUpdate is called from the calico client, whenever there
 // is a change to the svc_external_ips. The set of whitelisted external IP networks
 // will be updated to the new values specified in the default BGP configuration.
 // All of the routes advertised for each service will then be updated to reflect
 // this change.
-func (rg *routeGenerator) onBGPConfigurationUpdate(newExternalNets []*net.IPNet) {
+func (rg *routeGenerator) onExternalIPsUpdate(newExternalNets []string) {
+	rg.Lock()
+	// Check if it differs from our current configuration.
+	if reflect.DeepEqual(parseIPNets(newExternalNets), rg.externalIPNets) {
+		// No change, there's no work to do.
+		rg.Unlock()
+		return
+	}
 
+	// Determine the diff, and update the dataplane with it. First, find
+	// which CIDRs we should withdraw.
+	withdraws := []string{}
+	for _, existing := range rg.externalIPNets {
+		if !contains(newExternalNets, existing.String()) {
+			withdraws = append(withdraws, existing.String())
+		}
+	}
+	rg.externalIPNets = parseIPNets(newExternalNets)
+	rg.Unlock()
+
+	// Withdraw the old CIDRs and add the new.
```

Comment:
This is a behavior change - we weren't advertising the whole range before, but I think we should in order to properly implement Cluster type services.

Reply:
Just wonder if it would be worth splitting that out into a separate change?  These BGP changes are quite subtle, so when tracking down the history of a change like this it would be better to see it on its own and with a fuller discussion.